### DISCUSSION TITLE
Setting up of FPGA for pulsed measurements

### DISCUSSION THREAD

**Discussion by Vinay1786** [Ulm-IQO/qudi-core#32] [Q&A] [answered]
Dear All,

We have come across a couple of basic questions related to interfacing FPGA that we hope are good enough to be discussed here.

In the hardware module 'fast_counter_fpga_qo.py' :

(1) Why is the variable '_internal_clock_hz' is being set to 950 MHz even though the clock oscillator frequency is only 100 MHz for the FPGA version (Opal Kelly XEM6310) for which this code is written?

(2) For pulsed measurements, does the use of this FPGA hardware module ('fast_counter_fpga_qo.py') require an external clock/fast counter like the NI card? (the reason I am asking this question is because in a function that collects the hardware constraints, the code documentation says the function gets the 'hardware constraints of the fast counting device'. Since you mention about getting constraints from a fast counting device, I wonder whether you are just referring to the FPGA itself as a fast counting device or some other external hardware like NI card.)

If you could comment on these points that would be great!:-)

Thanks,
Vinaya

**Comment by Neverhorst** [comment, marked as answer]
Hi Vinaya,

before I answer your specific questions I would like to make sure we are on the same page regarding this FPGA qudi module.
In case you are thinking about buying one of those FPGA modules and just start measuring with qudi I have to disappoint you.
The qudi hardware module `fast_counter_fpga_qo.py` is just providing an interface to a specific FPGA-based hardware device. This device consists of more than just the bare FPGA module you can buy from Opal Kelly. It requires a custom made printed circuit board (PCB) as well as the matching FPGA configuration file or hardware description language (HDL) code that makes it behave as such a device.
Both, the PCB design as well as the FPGA HDL code, are proprietary and _not_ open source.

TL;DR: In order to use the mentioned qudi hardware module you would need proprietary hardware and firmware that is not available to the public.

Regarding your questions:

1. The 100 MHz clock is just the reference clock for the FPGA module. For the FPGA counter to work, it generates a multitude of phase-synchronous clocks with different frequencies from this reference clock via phase-locked-loops (PLL). The 950 MHz clock is the sampling clock for the digital input signals, i.e. the resolution of the generated time trace. All other clocks (including the 100 MHz) are purely for internal workings and not relevant for the user.
2. The "counting device" mentioned is the FPGA module itself. The function is returning the constraints characterizing the FPGA-based fast counter. The device just needs external synchronization trigger, gate trigger (active counting interval) and signal (e.g. APD) inputs.

**Reply by Vinay1786** [reply]
Hi Nikolas,

Many thanks for the kind reply and detailed answers to my queries. I appreciate it. Sorry, for being somewhat late to write a follow-up reply. 

Regarding the requirements to setup the FPGA module, you mentioned that we would need a custom made PCB and a matching FPGA config file. Firstly, by 'custom PCB', are you referring to the add-on 'breakout board' (https://opalkelly.com/products/brk7010/) that is available for purchase from OpalKelly or something we need to design ourselves ? 

Secondly, regarding the config file or HDL code, I read up from OK website that upon purchase of the FPGA hardware, the OK lets us download certain software. Is it the same you are talking about or something we must purchase from elsewhere?

Lastly, assuming that I have the custom PCB and a right config file/HDL, will it be straight forward to interface the FPGA module with rest of the hardware and run the experiments or do I need to work on the code to get things ready?

Thanks again for giving this opportunity to learn!

Regards,
Vinaya

**Reply by Neverhorst** [reply]
Hi Vinaya,

you are welcome.

1. No, it is custom-made as stated above. Meaning it is a PCB containing an active electronics design that is dependent on the FPGA HDL implementation. You can not build something like this without having the matching HDL code.
2. No, this is not a commercially available product. It has been designed by myself.
3. The PCB and the HDL code are proprietary and you can not simply buy this device.

So, to make it perfectly clear: There is no way for you to just buy a couple of things and make use of this hardware module.  
If you want a similar solution, you would need to start from scratch, buy an FPGA module (e.g. OK), implement HDL code to make it behave as a time tagger, design and manufacture a PCB to connect this FPGA to your lab devices and write a qudi hardware module to interface with it via software. The qudi hardware module is by far the smalles amount of work. I would highly recommend buying a commercial time tagging solution.

If you want to discuss the topic even further, please feel free to contact me directly via email.

**Reply by Vinay1786** [reply]
Hi Nikolas,

Many thanks again for the kind reply. That clears a lot of ambiguity. You are absolutely right, given the amount of time and efforts it takes to make an FPGA module to work, it makes sense to go for a commercial time-tagger. 

I would follow up with you via email for any further discussion on the topic.

Let me mark this question as answered for now. 

Thank you!
Best regards,
Vinaya
